clients/bazel-module-registry/src/main/kotlin/BazelModuleRegistryService.kt

```diff
@@ -0,0 +1,169 @@
+
+@Serializer(URI::class)
+object URISerializer : KSerializer<URI> {
+    override fun serialize(encoder: Encoder, value: URI) = encoder.encodeString(value.toString())
+    override fun deserialize(decoder: Decoder) = URI(decoder.decodeString())
+}
+
+/**
+ * Interface for a Bazel Module Registry, based on the directory structure of https://bcr.bazel.build
+ * and the git repository it is based on (https://github.com/bazelbuild/bazel-central-registry/).
+ */
+interface BazelModuleRegistryService {
+    companion object {
+        /**
+         * The JSON (de-)serialization object used by this service.
+         */
+        private val JSON = Json {
+            ignoreUnknownKeys = true
+            namingStrategy = JsonNamingStrategy.SnakeCase
+        }
+
+        /**
+         * The service uses the Bazel Central Registry by default.
+         */
+        const val DEFAULT_URL = "https://bcr.bazel.build"
+
+        /**
+         * Create a Bazel Module Registry service instance for communicating with a server running at the given [url],
+         * defaulting to the Bazel Central Registry, optionally with a pre-built OkHttp [client].
+         */
+        fun create(
+            url: String = DEFAULT_URL,
+            client: OkHttpClient? = null
+        ): BazelModuleRegistryService {
+            val bmrClient = client ?: OkHttpClient()
+
+            val contentType = "application/json".toMediaType()
+            val retrofit = Retrofit.Builder()
+                .client(bmrClient)
+                .baseUrl(url)
+                .addConverterFactory(JSON.asConverterFactory(contentType))
+                .build()
+
+            return retrofit.create(BazelModuleRegistryService::class.java)
+        }
+    }
+
+    @Serializable
+    data class ModuleMetadata(
+        @Serializable(URISerializer::class) val homepage: URI?,
+        val maintainers: List<ModuleMaintainer>? = null,
+        val repository: List<String>? = null,
+        val versions: List<String>,
+        // The key in the map is the version, the value the reason for yanking it.
+        val yankedVersions: Map<String, String>,
+    ) {
+        fun vcsInfo(): VcsInfo {
+            if (repository.isNullOrEmpty()) {
+                return VcsInfo.EMPTY
+            }
+
+            val repo = repository[0]
+            // From looking at all current values of this field on BCR, it looks like only the special value "github:"
+            // exists. Otherwise it's just the repo URL.
+            val url = if (repo.startsWith("github:")) {
+                val path = repo.substringAfter("github:")
+                "https://github.com/${path}"
+            } else {
+                repo
+            }
+
+            return VcsInfo(
+                type = VcsType.GIT,
+                url = url,
+                revision = "",
+                path = "",
+            )
+        }
+    }
+
+    @Serializable
+    data class ModuleMaintainer(
```

Comment:
Do you maybe have a link / reference to the upstream data model to add e.g. into KDoc for the data classes?

Reply:
I added links to example files on the Bazel Central Registry.

Reply:
Thanks!